[Site Isolation] Clicking on the body of a frame to focus it should blur the active element

aprotyas opened the pull request:
#### d207618b6bf4aafb72d8dd9ad1dc44d97aff565f
<pre>
[Site Isolation] Clicking on the body of a frame to focus it should blur the active element
<a href="https://bugs.webkit.org/show_bug.cgi?id=299543">https://bugs.webkit.org/show_bug.cgi?id=299543</a>
<a href="https://rdar.apple.com/157077681">rdar://157077681</a>

Reviewed by Wenson Hsieh.

With site isolation, a page&apos;s focus management is driven per web content
process for said process&apos;s local frame set. This means there are
multiple focus controllers when a page has cross-origin iframes loaded.
If a user clicks on a focusable element in some cross-origin iframe, the
focus controller belonging to the web content process where said iframe
is a local frame will grant focus to the element. However, some other
web content process may still think focus belongs to an element in one
of its local frames, because that is what its focus controller believes.

In this patch, we alleviate this issue by introducing the ability to
broadcast focused element changes across web content processes. This is
similar in spirit to 268570@main, which introduced the same ability
except for when a focused frame changed.

To do so, we hook into the Document::setFocusedElement() flow and teach
Chrome(Client) about other data required to correctly broadcast this
change, namely the focused local frame where the change happened and the
focus options.

More details in-line.

Test: http/tests/site-isolation/focus-click-navigation-activeElement.html

* LayoutTests/http/tests/site-isolation/focus-click-navigation-activeElement-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/focus-click-navigation-activeElement.html: Added.
* LayoutTests/http/tests/site-isolation/focus-tab-navigation-activeElement.html:
* LayoutTests/http/tests/site-isolation/resources/focus-click-navigation-iframe.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setFocusedElement):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):
(WebCore::Element::blur):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/cocoa/AutofillElements.cpp:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::focusedElementChanged):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setFocusedFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::setFocusedElement):

* Source/WebCore/page/FocusController.h:
We track a m_focusedFrameBeforeRemoteFocusBroadcast pointer, which
corresponds to the focused frame that got reset by a remote broadcast.
We then consult this pointer when resetting element focus because of a
remote broadcast, too.

* Source/WebCore/page/FocusControllerTypes.h: Added.
We take the liberty to pull out some types used by FocusController
for cross process coordination into this header.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedElementChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::focusPluginElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::focusedElementChanged):
(WebKit::WebChromeClient::focusedFrameChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::loadDidCommitInAnotherProcess):
(WebKit::WebFrame::commitProvisionalFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::elementWasFocusedInAnotherProcess):
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.mm:
(WebChromeClientIOS::focusedElementChanged):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::focusedElementChanged):

Canonical link: <a href="https://commits.webkit.org/300767@main">https://commits.webkit.org/300767@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/59d700a1901d0cf50f464beed5a5f6bed81ca581

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43454 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34150 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130515 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75888 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c2970946-4ad5-4fe7-aa76-c8cdc6ceec42) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125616 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44177 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52048 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94096 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62451 "Passed tests") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cca4e750-0392-46f5-a7ee-56296bd35e64) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126692 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35192 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110694 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74703 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/21a82eba-ca79-4cba-b88d-d9838cda371d) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/123095 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28854 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73996 "Built successfully") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104915 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29077 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133203 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50690 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38596 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102572 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51065 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106916 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102407 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47751 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25999 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47520 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19481 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50544 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56306 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50019 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53365 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51693 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/page/Chrome.cpp:
```diff
@@ -203,9 +205,9 @@ void Chrome::takeFocus(FocusDirection direction)
+void Chrome::focusedElementChanged(Element* element, Frame* frame, FocusOptions options, BroadcastFocusedElement broadcast)
 {
-    m_client->focusedElementChanged(element);
+    m_client->focusedElementChanged(element, frame, options, broadcast);
```
IIUC, `frame` should either be equal to the `element`'s frame if it's a local frame; otherwise, if it's a remote frame, the `element` must be null? (If that's accurate, it might be good to add an assertion here to enforce that invariant).

aprotyas replied:
That's correct, I added assertions in a couple of places! (`Chrome:: focusedElementChanged()` and `FocusController::setFocusedElement()`)

whsieh commented on Source/WebCore/page/FocusController.h:
```diff
@@ -144,6 +134,7 @@ class FocusController final : public CanMakeCheckedPtr<FocusController> {
 
     WeakRef<Page> m_page;
     WeakPtr<Frame> m_focusedFrame;
+    WeakPtr<Frame> m_focusedFrameBeforeRemoteFocusBroadcast;
```
It looks like `m_focusedFrameBeforeRemoteFocusBroadcast` can only ever be a `WebCore::LocalFrame` — if so, we could change this to `WeakPtr<LocalFrame>` and remove the `dynamicDowncast<LocalFrame>` above.

aprotyas replied:
Done!

aprotyas committed:
[Site Isolation] Clicking on the body of a frame to focus it should blur the active element
https://bugs.webkit.org/show_bug.cgi?id=299543
rdar://157077681

Reviewed by Wenson Hsieh.

With site isolation, a page's focus management is driven per web content
process for said process's local frame set. This means there are
multiple focus controllers when a page has cross-origin iframes loaded.
If a user clicks on a focusable element in some cross-origin iframe, the
focus controller belonging to the web content process where said iframe
is a local frame will grant focus to the element. However, some other
web content process may still think focus belongs to an element in one
of its local frames, because that is what its focus controller believes.

In this patch, we alleviate this issue by introducing the ability to
broadcast focused element changes across web content processes. This is
similar in spirit to 268570@main, which introduced the same ability
except for when a focused frame changed.

To do so, we hook into the Document::setFocusedElement() flow and teach
Chrome(Client) about other data required to correctly broadcast this
change, namely the focused local frame where the change happened and the
focus options.

More details in-line.

Test: http/tests/site-isolation/focus-click-navigation-activeElement.html

* LayoutTests/http/tests/site-isolation/focus-click-navigation-activeElement-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/focus-click-navigation-activeElement.html: Added.
* LayoutTests/http/tests/site-isolation/focus-tab-navigation-activeElement.html:
* LayoutTests/http/tests/site-isolation/resources/focus-click-navigation-iframe.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setFocusedElement):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus):
(WebCore::Element::blur):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::setFocusedElementIfNeeded):
* Source/WebCore/editing/cocoa/AutofillElements.cpp:
* Source/WebCore/loader/EmptyClients.h:
* Source/WebCore/page/Chrome.cpp:
(WebCore::Chrome::focusedElementChanged):
* Source/WebCore/page/Chrome.h:
* Source/WebCore/page/ChromeClient.h:
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::dispatchMouseEvent):
* Source/WebCore/page/FocusController.cpp:
(WebCore::FocusController::setFocusedFrame):
(WebCore::FocusController::findFocusableElementDescendingIntoSubframes):
(WebCore::FocusController::setFocusedElement):

* Source/WebCore/page/FocusController.h:
We track a m_focusedFrameBeforeRemoteFocusBroadcast pointer, which
corresponds to the focused frame that got reset by a remote broadcast.
We then consult this pointer when resetting element focus because of a
remote broadcast, too.

* Source/WebCore/page/FocusControllerTypes.h: Added.
We take the liberty to pull out some types used by FocusController
for cross process coordination into this header.

* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/WebFrameProxy.cpp:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::focusedElementChanged):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::focusPluginElement):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::focusedElementChanged):
(WebKit::WebChromeClient::focusedFrameChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::loadDidCommitInAnotherProcess):
(WebKit::WebFrame::commitProvisionalFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::elementWasFocusedInAnotherProcess):
(WebKit::WebPage::frameWasFocusedInAnotherProcess):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::focusTextInputContextAndPlaceCaret):
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.h:
* Source/WebKitLegacy/ios/WebCoreSupport/WebChromeClientIOS.mm:
(WebChromeClientIOS::focusedElementChanged):
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebChromeClient.mm:
(WebChromeClient::focusedElementChanged):

Canonical link: https://commits.webkit.org/300767@main
